### PR TITLE
  London Class 7 -Negin Dn - Javascript core 1 - Week 3

### PR DESCRIPTION
**Pull request by NeginDn**
# Your Details

- Your Name:
- Your City:
- Your Slack Name:

# Homework Details

- Module:
- Week:

### CODE REVIEW

File: 1-exercises/A-array-find/exercise.js [hunk 1]
```diff
@@ -17,7 +17,11 @@ var names = [
   "Ahmed",
 ];
 
-var longNameThatStartsWithA = findLongNameThatStartsWithA(names);
+function findLongNameThatStartsWithA(name) {
```
Comment (louisechow) [inline 1-exercises/A-array-find/exercise.js:20]: This all works 👏. I'm going to make some suggestions that I think will make your code cleaner.

It's generally good practice to name a [predicate function](https://github.com/CodeYourFuture/JavaScript-Core-1-Coursework-Week2/tree/main/exercises/D-predicates) (a function that returns `true` or `false`) starting with a verb that can be answered with yes or no, such as `is`, `can`, `should`.
`isPostiveNumber` -> yes/no
`canUpdateStatus` -> yes/no
`shouldDeleteAccount` -> yes/no

I would suggest following this convention for the predicate function here. 

But then you might ask... why does the exercise mention a function called `findLongNameThatStartsWithA`?  I think that indicates you should create a new function with that name that contained the `find` logic there.
```
function findLongNameThatStartsWithA(names) {
  return names.find((name) => isLongNameThatStartsWithA(name));
}

var longNameThatStartsWithA = findLongNameThatStartsWithA(names);
```

File: 1-exercises/B-array-some/exercise.js [hunk 2]
```diff
@@ -15,7 +15,15 @@ var pairsByIndex = [[0, 3], [1, 2], [2, 1], null, [3, 0]];
 var students = ["Islam", "Lesley", "Harun", "Rukmini"];
 var mentors = ["Daniel", "Irina", "Mozafar", "Luke"];
 
+function isNull(indexes) {
```
Comment (louisechow) [inline 1-exercises/B-array-some/exercise.js:18]: This is great naming for this predicate function.

File: 1-exercises/B-array-some/exercise.js [hunk 3]
```diff
@@ -15,7 +15,15 @@ var pairsByIndex = [[0, 3], [1, 2], [2, 1], null, [3, 0]];
...
+  return indexes === null;
+}
+
 var pairs = pairsByIndex.map(function (indexes) {
+  if (isNull) {
```
Comment (louisechow) [inline 1-exercises/B-array-some/exercise.js:23]: 🤔 You're calling the `isNull` function here directly, but you aren't passing it any parameters. If you don't pass in a parameter, `indexes` on line 18 will always be `null` so when you run your program, it will always exit, even if it is given correct data.

What you want, is to use an array method that will allow you to pass in your function `isNull` as an argument and will return true if there is a null in the array, then you can call `process.exit(1)`. Which array method do you think you need to use?

File: 1-exercises/C-array-every/exercise.js [hunk 4]
```diff
@@ -5,7 +5,10 @@
 var students = ["Omar", "Austine", "Dany", "Swathi", "Lesley", "Rukmini"];
 var group = ["Austine", "Dany", "Swathi", "Daniel"];
 
-var groupIsOnlyStudents; // complete this statement
+function everyoneIsStudents(groupName) {
```
Comment (louisechow) [inline 1-exercises/C-array-every/exercise.js:8]: Good job!

Only suggestion is to follow the convention I mentioned earlier for naming predicate functions.

File: 1-exercises/D-array-filter/exercise.js [hunk 5]
```diff
@@ -8,7 +8,9 @@
 
 var pairsByIndexRaw = [[0, 3], [1, 2], [2, 1], null, [1], false, "whoops"];
 
-var pairsByIndex; // Complete this statement
+var pairsByIndex = pairsByIndexRaw.filter((indexes) => {
+  return indexes !== "boolean" && indexes !== null && indexes.length === 2;
```
Comment (louisechow) [inline 1-exercises/D-array-filter/exercise.js:12]: Why did you add `indexes !== "boolean"`? What does this do?

File: 1-exercises/E-array-map/exercise.js [hunk 6]
```diff
@@ -2,3 +2,31 @@
...
+  return number3 * 100;
+});
+
+console.log("Case 3: " + multiplied3);
+
+//4
+let multiplied4 = numbers.map((number4) => {
+  return number4 * 100;
+});
+
```
Comment (louisechow) [inline 1-exercises/E-array-map/exercise.js:31]: You can go one level more and get rid of the `{}` and the `return` statement.

File: 1-exercises/F-array-forEach/exercise.js [hunk 7]
```diff
@@ -8,6 +8,23 @@
...
+  } else {
+    return number;
+  }
+}
+
+function print(number) {
+  console.log(number);
+}
+
+arr.map(multiples).forEach(print);
```
Comment (louisechow) [inline 1-exercises/F-array-forEach/exercise.js:27]: 👍

File: 1-exercises/H-array-methods-2/exercise2.js [hunk 8]
```diff
@@ -7,7 +7,9 @@
   Tip: use the string method .split() and the array method .join()
 */
 
-function capitalise(str) {}
+function capitalise(str) {
```
Comment (louisechow) [inline 1-exercises/H-array-methods-2/exercise2.js:10]: Nice!

File: 1-exercises/J-string-substring/exercise2.js [hunk 9]
```diff
@@ -14,11 +14,11 @@ let names = [
   "Arron Graham",
 ];
 
-names[0] = names[0].substring();
-names[1] = names[1].substring();
-names[2] = names[2].substring();
-names[3] = names[3].substring();
-names[4] = names[4].substring();
```
Comment (louisechow) [inline 1-exercises/J-string-substring/exercise2.js:21]: 👍

File: 2-mandatory/1-oxygen-levels.js [hunk 10]
```diff
@@ -3,15 +3,20 @@
...
     To be safe, they need to land on the first unnamed planet that has Oxygen levels between 19.5% and 23.5%.
-
     Write a function that finds the oxygen level of the first safe planet - Oxygen between 19.5% and 23.5%
-
     Some string methods that might help you here are .replace() and .substring(). 
 */
 
-function findSafeOxygenLevel() {}
+function findSafeOxygenLevel(planetOxygenLevel) {
+  const OxygenInNumber = planetOxygenLevel.map((number) =>
```
Comment (louisechow) [inline 2-mandatory/1-oxygen-levels.js:12]: Why did you use `const` here? 🤔 

Usually we only use `const` if we are setting a variable that should never change. E.g. `const PI = 3.14` or `const TAX_RATE_PERCENT = 20`. 

Also, the convention is usually to use all uppercase when naming a constant variable e.g. `OXYGEN_IN_NUMBER`. If you make only the first letter uppercase, it usually means that is an Object.  

https://www.freecodecamp.org/news/javascript-naming-conventions-dos-and-don-ts-99c0e2fdd78a/

File: 2-mandatory/1-oxygen-levels.js [hunk 11]
```diff
@@ -3,15 +3,20 @@
...
+    number.replace("%", "")
+  );
+  const isOnRange = OxygenInNumber.find(
```
Comment (louisechow) [inline 2-mandatory/1-oxygen-levels.js:15]: I think you could name `isOnRange` to be more descriptive. 
The `find` method will return the first value that matches, not `true` or `false`, so it's not a good scenario to start the variable name with `is`.

A better name would be `safeOxygenLevel`. Then you would have a number assigned that would be a safe oxygen level.

File: 2-mandatory/2-bush-berries.js [hunk 12]
```diff
@@ -21,8 +21,14 @@
   Let's first look at an example that will teach you how to use these methods.
 */
 
-function isBushSafe() {}
-
+function isBushSafe(bush) {
+  const checkBurry = (colour) => colour === "pink";
```
Comment (louisechow) [inline 2-mandatory/2-bush-berries.js:25]: I would use `let` here instead of `const`. Also try to follow the convention for naming a predicate function staring with `is`, `can`, `should` etc.

File: 2-mandatory/3-space-colonies.js [hunk 13]
```diff
@@ -15,8 +15,11 @@
   
 */
 
-function getSettlers() {}
-
+function getSettlers(voyagers) {
+
```
Comment (louisechow) [inline 2-mandatory/3-space-colonies.js:19]: 👍 Great logic here!

I don't think you need to assign this to a variable (particularly not a `const`), you can just return the statement directly.

File: 2-mandatory/4-eligible-students.js [hunk 14]
```diff
@@ -7,8 +7,9 @@
   - Returns an array containing only the names of the who have attended AT LEAST 8 classes
 */
 
-function getEligibleStudents() {
-
```
Comment (louisechow) [inline 2-mandatory/4-eligible-students.js:11]: Great logic here! 👍 

Here is a good scenario where using `const` will make your code more readable.
If we create some constants
```
const MINIMUM_CLASSES = 8;
const NAME_INDEX = 0;
const ATTENDANCE_INDEX = 1;
```
Then you could rewrite line 11 with the constants (and some method chaining) to make it easy to read what is happening.
```
let eligibleStudents = attendance.filter(student => student[ATTENDANCE_INDEX] > = MINIMUM_CLASSES).map(student => student[NAME_INDEX]);
```

File: 2-mandatory/5-journey-planner.js [hunk 15]
```diff
@@ -81,7 +83,13 @@ function getTransportModes() {}
     
   Hint: Use the corresponding array method to decide if an element is member of an array.
 */
-function isAccessibleByTransportMode() {}
+function isAccessibleByTransportMode(transportList, transportMode) {
+  if (transportList.includes(transportMode)) {
```
Comment (louisechow) [inline 2-mandatory/5-journey-planner.js:87]: The `includes` method will return `true` or `false` for you. So you don't need the `if...else...` condition.

File: 2-mandatory/5-journey-planner.js [hunk 16]
```diff
@@ -122,6 +132,9 @@ function getLocationName() {}
   Advanced challange: try to use arrow function when invoking an array method.
 */
 function journeyPlanner(locations, transportMode) {
+  return locations
+    .filter((locations) => locations.includes(transportMode))
+    .map((locations) => locations[0]);
   // Implement the function body
```
Comment (louisechow) [inline 2-mandatory/5-journey-planner.js:138]: Your logic is good here!

To make it more readable, you can replace some of your code with the functions that you defined earlier in the exercise.
`locations.includes(transportMode)` => `isAccessibleByTransportMode(getTransportModes(locations)`, transportMode)
`locations[0]` => `getLocationName(locations)`

File: 2-mandatory/6-lane-names.js [hunk 17]
```diff
@@ -6,7 +6,12 @@
   HINT: string and array methods that could be helpful (indexOf, filter)
 */
```
Comment (louisechow) [inline 2-mandatory/6-lane-names.js:8]: Great logic! Similar comments about not needing to assign to a variable. You can just return the statement directly.

File: 2-mandatory/7-password-validator.js [hunk 18]
```diff
@@ -25,51 +25,63 @@ PasswordValidationResult=  [false, false, false, false, true]
 */
 
 function validatePasswords(passwords) {
+  const acceptedPassword = passwords.map(
```
Comment (louisechow) [inline 2-mandatory/7-password-validator.js:28]: good job 👍

**Comment by louisechow** [comment]
Great job Negin! 🌟 You're very good at choosing meaningful names for your variables and you often choose very simple and elegant solutions to the exercises.

Most of my comments for improvements are about:
- Using `const` at the right times. Most of the time you should be using `let`.
- Naming of your functions - Follow convention to name predicate functions starting with a verb like is, can, should etc. For other functions, try to choose a name that describes what the function is doing; is it returning a value (maybe start it with `get`), is it returning a list or a single item (choose the right word, either singular or plural).

https://syllabus.codeyourfuture.io/guides/code-style-guide/#think-of-good-names-for-your-variables
https://www.freecodecamp.org/news/javascript-naming-conventions-dos-and-don-ts-99c0e2fdd78a/

File: 1-exercises/D-array-filter/exercise.js (hunk 5 above)
Comment (NeginDn) [reply 1-exercises/D-array-filter/exercise.js:12]: The goal was to check if the array contains boolian(which shouldn't), shouldn't I have done that?

File: 1-exercises/E-array-map/exercise.js (hunk 6 above)
Comment (NeginDn) [reply 1-exercises/E-array-map/exercise.js:31]: Done, Perfect!  I didn't know this

File: 2-mandatory/2-bush-berries.js (hunk 12 above)
Comment (NeginDn) [reply 2-mandatory/2-bush-berries.js:25]: About the naming I understand now how it works but since the function name in test was isBushSafe I used that. Should I have two different function here?

File: 2-mandatory/4-eligible-students.js (hunk 14 above)
Comment (NeginDn) [reply 2-mandatory/4-eligible-students.js:11]: Do we name const variables with all capital letters? And also I found chaining very confusing.I did read some articles but made it worse. Is there an easy to understand video I can watch? If not is it wrong not to use them just yet?

File: 2-mandatory/5-journey-planner.js (hunk 16 above)
Comment (NeginDn) [reply 2-mandatory/5-journey-planner.js:138]: I am having trouble understanding it. Maybe I can ask you about it in our last session?

**Comment by NeginDn** [comment]
@louisechow Thank you so much for all of the comments. I've learned a lot form your feedback and will try to use them in my next homeworks. I really appreciate it
